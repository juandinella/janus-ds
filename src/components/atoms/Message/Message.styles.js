import { css } from 'styled-components'
import getContrastRatio from '../../../utils/getContrastRatio'

const getReadableTextColor = (backgroundColor) => {
  const black = '171717'
  const white = 'FFFFFF'

  const contrastWithBlack = getContrastRatio(backgroundColor, black)
  const contrastWithWhite = getContrastRatio(backgroundColor, white)

  if (contrastWithBlack > contrastWithWhite) {
    return `#${black}`
  } else {
    return `#${white}`
  }
}

export default css`
  padding: var(--spacing-xxs) var(--spacing-xs);
  font-size: var(--font-size-sm);
  max-width: 305px;
  white-space: pre-line;
  border-radius: var(--border-radius-md);
  line-height: 1.5;

  ${(props) =>
    props.variant === 'incoming' &&
    css`
      color: var(--color-neutral-900);
      background-color: var(--color-neutral-100);
    `}

  ${(props) =>
    props.variant === 'outgoing' &&
    css`
      color: ${getReadableTextColor(props.backgroundColor)};
      background-color: ${props.backgroundColor};
    `}

  ${(props) =>
    props.loading &&
    css`
      display: flex;
      gap: 5px;

      span {
        width: 10px;
        height: 10px;
        border-radius: var(--border-radius-round);
        animation: dots-animation 1.5s infinite ease-in-out;

        @media screen and (prefers-reduced-motion) {
          animation: none;
        }

        &:nth-child(1) {
          animation-delay: calc(300ms * 1);
        }
        &:nth-child(2) {
          animation-delay: calc(300ms * 2);
        }
        &:nth-child(3) {
          animation-delay: calc(300ms * 3);
        }

        @keyframes dots-animation {
          0% {
            background-color: var(--color-neutral-100);
          }
          20% {
            background-color: var(--color-neutral-400);
          }
          44% {
            background-color: var(--color-neutral-700);
          }
        }
      }
    `}
`
