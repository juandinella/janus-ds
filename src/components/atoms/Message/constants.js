export const options = {
  variants: ['incoming', 'outgoing'],
}
