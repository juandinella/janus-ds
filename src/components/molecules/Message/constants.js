export const options = {
  directions: ['incoming', 'outgoing'],
  positions: ['single', 'first', 'middle', 'last']
}
