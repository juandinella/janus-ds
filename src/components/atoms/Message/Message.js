import isPropValid from '@emotion/is-prop-valid'
import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'
import { options } from './constants'
import styles from './Message.styles'

/**
 * Message
 */

const shouldForwardProp = (prop) => isPropValid(prop)
const StyledMessage = styled.div.withConfig({ shouldForwardProp })`
  ${styles}
`

const Message = React.forwardRef(
  ({ id, variant, children, backgroundColor = '#171717' }, ref) => {
    return (
      <StyledMessage
        ref={ref}
        id={id}
        data-testid={id}
        variant={variant}
        backgroundColor={backgroundColor}
      >
        <span data-testid={`${id}-children`}>{children}</span>
      </StyledMessage>
    )
  },
)

Message.displayName = 'Message'

Message.propTypes = {
  id: PropTypes.string,
  children: PropTypes.node.isRequired,
  variant: PropTypes.oneOf(options.variants),
  backgroundColor: PropTypes.string,
}

export default Message
