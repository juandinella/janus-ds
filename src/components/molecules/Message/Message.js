import isPropValid from '@emotion/is-prop-valid'
import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'
import { options } from './constants'
import styles from './Message.styles'

import Avatar from '../../atoms/Avatar'
import Bubble from '../../atoms/Bubble'
import Spacer from '../../layout/Spacer'

/**
 * Message
 */

const shouldForwardProp = (prop) => isPropValid(prop)
const StyledMessage = styled.div.withConfig({shouldForwardProp})`${styles}`

const Message = ({ id, direction, position, children }) => {
  return (
    <StyledMessage id={id} direction={direction} position={position}>
      {direction === 'incoming' &&
        <>
          <Avatar size='sm'>GR</Avatar>
          <Spacer size='8' />
        </>
      }
      <Bubble variant={direction}>{children}</Bubble>
    </StyledMessage>
  )
}

Message.displayName = 'Message'

Message.propTypes = {
  children: PropTypes.node.isRequired,
  direction: PropTypes.oneOf(options.directions),
}

Message.defaultProps = {}

export default Message
