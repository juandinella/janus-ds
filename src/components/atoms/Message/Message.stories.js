import { options } from './constants'
import Message from './Message'

import {
  getTemplate,
  getListTemplate,
  getOptionsArgTypes,
} from '../../../helpers/storybook'

const Template = getTemplate(Message)
const ListTemplate = getListTemplate(Message)

export default {
  title: 'Atoms/Message',
  component: Message,
  args: {
    children: 'Soy un mensaje de prueba',
    variant: 'incoming',
    id: 'test',
    backgroundColor: '#171717',
  },
  argTypes: {
    ...getOptionsArgTypes(options.variants, 'variant'),
    backgroundColor: {
      control: { type: 'color' },
    },
  },
}

export const Default = Template.bind({})

export const Outgoing = {
  args: {
    backgroundColor: '#171717',
    variant: 'outgoing',
  },
}

export const Variants = ListTemplate.bind({})
Variants.args = { items: options.variants.map((variant) => ({ variant })) }

export const Loading = { args: { loading: true } }
