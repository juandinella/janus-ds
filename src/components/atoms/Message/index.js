import Message from './Message'

export default Message
export { options } from './constants'
