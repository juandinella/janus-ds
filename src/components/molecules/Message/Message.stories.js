import { options } from './constants'
import Message from './Message'

import {
  getTemplate,
  getListTemplate,
  getOptionsArgTypes,
} from '../../../helpers/storybook'

const Template = getTemplate(Message)
const ListTemplate = getListTemplate(Message)

export default {
  title: 'Molecules/Message',
  component: Message,
  args: {
    children: 'Soy un mensaje de prueba',
    direction: 'incoming'
  },
  argTypes: {
    ...getOptionsArgTypes(options.directions, 'direction'),
    ...getOptionsArgTypes(options.positions, 'position'),
  },
}

export const Default = Template.bind({})

export const SingleIncoming = { args: { direction: 'incoming', position: 'single' } }
export const SingleOutgoing = { args: { direction: 'outgoing', position: 'single' } }

