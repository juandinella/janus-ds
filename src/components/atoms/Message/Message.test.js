import initStoryshots from '@storybook/addon-storyshots'
import { render, screen, cleanup } from '@testing-library/react'
import React from 'react'
import Message from './Message'

initStoryshots({
  storyKindRegex: /^Atoms\/Message$/,
})

const componentProps = {
  id: 'unique_id',
  children: 'Soy un mensaje de prueba',
  variant: 'aiMessage',
}
afterEach(cleanup)

describe('<Message />', () => {
  test('should render component correctly', () => {
    render(<Message {...componentProps} />)
    const msg = screen.getByTestId(componentProps.id)
    expect(msg).toBeInTheDocument()
  })

  test('should render children correctly', () => {
    render(<Message {...componentProps} />)
    const childrenElement = screen.getByTestId(`${componentProps.id}-children`)
    expect(childrenElement.textContent).toEqual(componentProps.children)
  })
})
