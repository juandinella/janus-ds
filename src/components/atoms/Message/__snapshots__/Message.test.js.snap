// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Atoms/Message Default 1`] = `
<div
  style={
    {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
      "position": "relative",
    }
  }
>
  <div
    className="sc-1kelln5-0 exuiBe"
    data-testid="test"
    id="test"
  >
    <span
      data-testid="test-children"
    >
      Soy un mensaje de prueba
    </span>
  </div>
</div>
`;

exports[`Storyshots Atoms/Message Loading 1`] = `
<div
  style={
    {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
      "position": "relative",
    }
  }
>
  <div
    className="sc-1kelln5-0 guGDmx"
    data-testid="test"
    id="test"
    loading={true}
  >
    <span />
    <span />
    <span />
  </div>
</div>
`;

exports[`Storyshots Atoms/Message Outgoing 1`] = `
<div
  style={
    {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
      "position": "relative",
    }
  }
>
  <div
    className="sc-1kelln5-0 gsOApI"
    data-testid="test"
    id="test"
  >
    <span
      data-testid="test-children"
    >
      Soy un mensaje de prueba
    </span>
  </div>
</div>
`;

exports[`Storyshots Atoms/Message Variants 1`] = `
<div
  style={
    {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
      "position": "relative",
    }
  }
>
  <div
    className="sc-1kelln5-0 exuiBe"
    data-testid="test"
    id="test"
  >
    <span
      data-testid="test-children"
    >
      Soy un mensaje de prueba
    </span>
  </div>
  <div
    className="sc-1kelln5-0 gsOApI"
    data-testid="test"
    id="test"
  >
    <span
      data-testid="test-children"
    >
      Soy un mensaje de prueba
    </span>
  </div>
</div>
`;
